analyzers/tests/SonarAnalyzer.UnitTest/Extensions/AssignmentExpressionSyntaxExtensionsTests.cs

```diff
@@ -0,0 +1,364 @@
+            var mapping = ParseAssignmentExpression(code).MapAssignmentArguments();
+            mapping.Should().BeEquivalentTo(new[]
+                {
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "a" } },
+                        Right = new { Token = new { Text = "1" } },
+                    },
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "b" } },
+                        Right = new { Token = new { Text = "2" } },
+                    },
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "c" } },
+                        Right = new { Token = new { Text = "3" } },
+                    },
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "d" } },
+                        Right = new { Token = new { Text = "4" } },
+                    },
+                });
+        }
+
+        [TestMethod]
+        public void AssignmentExpressionSyntaxExtensions_MisalignedDeconstructionAssignment()
+        {
+            var code = "var (a, (b, c, d)) =  (1, (2, 3), 4);";
+            var mapping = ParseAssignmentExpression(code).MapAssignmentArguments();
+            mapping.Should().BeEquivalentTo(new[]
+                {
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "a" } },
+                        Right = new { Token = new { Text = "1" } },
+                    },
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "b" } },
+                        Right = new { Token = new { Text = "2" } },
+                    },
+                    new
+                    {
+                        Left = new { Identifier = new { Text = "c" } },
+                        Right = new { Token = new { Text = "3" } },
+                    },
+                });
+        }
+
+        [TestMethod]
+        public void AssignmentExpressionSyntaxExtensions_MixedAssignment()
+        {
+            var code = "int a; (a, var b) =  (1, 2);";
+            var mapping = ParseAssignmentExpression(code).MapAssignmentArguments();
+            mapping.Should().HaveCount(2);
+            mapping[0].Should().BeEquivalentTo(new
+            {
+                Left = new { Identifier = new { Text = "a" } },
+                Right = new { Token = new { Text = "1" } },
+            });
+            mapping[1].Should().BeEquivalentTo(new
+            {
+                Left = new { Designation = new { Identifier = new { Text = "b" } } },
+                Right = new { Token = new { Text = "2" } },
+            });
+        }
+
+        [TestMethod]
+        public void AssignmentExpressionSyntaxExtensions_MisalignedLeft()
+        {
+            var code = "(var x, var y) = (1, 2, 3);";
```

Comment:
Same here, basically it would be great to detect all the misalignments and return a clear indication so the consumer does not assume that everything was correct.

Reply:
I see three options:

1. Return the valid mappings and ignore the remains (current behavior)
1. Return an empty array if anything does not fit
1. Return a single element array mapping tuple to tuple.
1. Throw InvalidOperationException

I think 1) and 3) are the best options. What do you think?

Reply:
> I see four options:
> 
> 1. Return the valid mappings and ignore the remains (current behavior)
**_With this option the consumer of the method does not know that the syntax is invalid and can create some noise and raise FPs_**
> 2. Return an empty array if anything does not fit
**_This will be a clear indication that something is not correct with the syntax and the consumer can try to parser it on its own if needed._**
> 3. Return a single element array mapping tuple to tuple.
**_The user will have again a clear indication that something is not correct with the syntax and can either parse on its own or can use the right and left side of the assignment for further processing._**
> 4. Throw InvalidOperationException
**_I would definitely suggest to now throw exceptions as the analyzers are used inside the IDE and it is a typical scenario to be run for invalid syntax_**


All in all I would go with either option 2 or 3. Up to you to decide.

Reply:
I went for 3)